### PR TITLE
drivers: spi: gd32: Add support DMA transfer 

### PR DESCRIPTION
**Pull request by soburi**
Add supporting DMA-based transfer for GD32 SPI.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

### CODE REVIEW

**Review by mbolivar-nordic — COMMENTED** [review]
@teburd -- I'm trying to take a step back from reviewing subsystem-specific DT changes in favor of asking those subsystem maintainers to take this on. I'm going to leave this up to you -- please feel free to ping if you have any specific questions

**Comment by nandojve** [comment]
Hi @soburi ,

You can rebase since #47501 got merge.

**Comment by soburi** [comment]
Hi, @nandojve , @cameled
Thank you for your continued support.
As with #47501, please review it.

File: dts/bindings/spi/gd,gd32-spi.yaml
Comment (galak) [inline dts/bindings/spi/gd,gd32-spi.yaml:28]: Why isn't this using a `dmas` property like dts/bindings/spi/atmel,sam0-spi.yaml

**Review by galak — DISMISSED** [review]
Use standard `dmas` property.

**Comment by cameled** [comment]
@soburi In general, gd32 evaluate boards have a spi flash. I have tried to run dma on that spi flash from gd32f450i_eval board, but looks there have something blocked.  I will try to dig out tomorrow. Anyway, this is an impressive work.

File: dts/bindings/spi/gd,gd32-spi.yaml
Comment (soburi) [reply dts/bindings/spi/gd,gd32-spi.yaml:28]: I will apply your pointing out.
dma-gd32.c need fix prior than spi,
so I create new PR #48756.
Could you please review also it?

**Review by cameled — COMMENTED** [review]
@soburi , Please add below patch to support run `samples/drivers/spi_flash` sample with DMA on `gd32f450i_eval` board.

```
diff --git a/boards/arm/gd32f450i_eval/gd32f450i_eval.dts b/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
index cf90ba9b39..2ca2443c87 100644
--- a/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
+++ b/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
@@ -144,11 +144,19 @@
 	};
 };
 
+&dma1 {
+	status = "okay";
+};
+
 &spi5 {
 	status = "okay";
 	pinctrl-0 = <&spi5_default>;
 	pinctrl-names = "default";
 	cs-gpios = <&gpioi 8 GPIO_ACTIVE_LOW>;
+	dmas = <&dma1 5>, <&dma1 6>;
+	dma-names = "tx", "rx";
+	tx-dma-peripheral = <1>;
+	rx-dma-peripheral = <1>;
 
 	nor_flash: gd25q16@0 {
 		compatible ="jedec,spi-nor";
diff --git a/samples/drivers/spi_flash/boards/gd32f450i_eval.conf b/samples/drivers/spi_flash/boards/gd32f450i_eval.conf
index b5aa8903ef..6d3d1db11f 100644
--- a/samples/drivers/spi_flash/boards/gd32f450i_eval.conf
+++ b/samples/drivers/spi_flash/boards/gd32f450i_eval.conf
@@ -5,3 +5,4 @@
 #
 
 CONFIG_SPI_NOR=y
+CONFIG_SPI_GD32_DMA=y
```

File: dts/bindings/spi/gd,gd32-spi.yaml
Comment (cameled) [inline dts/bindings/spi/gd,gd32-spi.yaml:38]: Please refer https://github.com/zephyrproject-rtos/zephyr/pull/42215#discussion_r827790806. 

We need to provide a simple way for the user to select the peripheral easily.

**Comment by cameled** [comment]
Please consider to split the original commit with `drivers` and `dts`.

File: dts/bindings/spi/gd,gd32-spi.yaml
Comment (cameled) [reply dts/bindings/spi/gd,gd32-spi.yaml:38]: This can be done in another PR, not related to this.

**Comment by nandojve** [comment]
Hi @soburi ,

Do you think can finish in the next 2 days? Sep. 9th is last day to get it in otherwise only for next release.

**Comment by soburi** [comment]
Hi, @nandojve 

> Hi @soburi ,
> 
> Do you think can finish in the next 2 days? Sep. 9th is last day to get it in otherwise only for next release.

I can rewrite the source with in 2 days, but #48756 discussion may not solve.
I will prepare, but I think it will be difficult for the above reasons.

**Comment by soburi** [comment]
Mark as draft until #48756 solved.

File: dts/bindings/dma/gd,gd32-dma.yaml
Comment (cameled) [inline dts/bindings/dma/gd,gd32-dma.yaml:73]: From GD32F4xx User Manual, slot (PERIEN btis) can be determined by Channel and Peripheral. For example, DMA0 Channel 0 and SPI2_RX only have one PERIEN, which value is 0.

After that, we can define some helper macro, like DMA0_CH0_SPI2_RX, user can easily to select them.  helper macro implement can refer `include/zephyr/dt-bindings/reset/gd32f3x0.h`.

File: dts/bindings/dma/gd,gd32-dma.yaml
Comment (cameled) [inline dts/bindings/dma/gd,gd32-dma.yaml:18]: Please fix the alignment.

File: boards/arm/gd32f450i_eval/gd32f450i_eval.dts
Comment (cameled) [inline boards/arm/gd32f450i_eval/gd32f450i_eval.dts:155]: Patch to sync with new dma cells setting.
```
diff --git a/boards/arm/gd32f450i_eval/gd32f450i_eval.dts b/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
index ba87bb8e1f..a9614525d2 100644
--- a/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
+++ b/boards/arm/gd32f450i_eval/gd32f450i_eval.dts
@@ -152,10 +152,8 @@
 	pinctrl-0 = <&spi5_default>;
 	pinctrl-names = "default";
 	cs-gpios = <&gpioi 8 GPIO_ACTIVE_LOW>;
-	dmas = <&dma1 5>, <&dma1 6>;
+	dmas = <&dma1 5 1 0 0>, <&dma1 6 1 0 0>;
 	dma-names = "tx", "rx";
-	tx-dma-peripheral = <1>;
-	rx-dma-peripheral = <1>;
 
 	nor_flash: gd25q16@0 {
 		compatible ="jedec,spi-nor";
```

**Comment by nandojve** [comment]
> Hi, @nandojve
> 
> > Hi @soburi ,
> > Do you think can finish in the next 2 days? Sep. 9th is last day to get it in otherwise only for next release.
> 
> I can rewrite the source with in 2 days, but #48756 discussion may not solve. I will prepare, but I think it will be difficult for the above reasons.

Hi @soburi ,

I think it is not necessary to be in rush. We already made very good progress for this release.

**Comment by soburi** [comment]
> This pull request has been marked as stale because it has been open (more than) 60 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this pull request will automatically be closed in 14 days. Note, that you can always re-open a closed pull request at any time.

Rebase to solve the stall.
It is still waiting  #48756 discussion result.

**Comment by nandojve** [comment]
> > This pull request has been marked as stale because it has been open (more than) 60 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this pull request will automatically be closed in 14 days. Note, that you can always re-open a closed pull request at any time.
> 
> Rebase to solve the stall. It is still waiting #48756 discussion result.

Hi @soburi,
In this case, ask feedback for @person to clarify missing points.
Remember that DMA is quite complex drivers and it must have tests to cover it. SPI have their own tests with DMA  at `tests/drivers/spi/spi_loopback`.

File: dts/bindings/dma/gd,gd32-dma.yaml
Comment (soburi) [reply dts/bindings/dma/gd,gd32-dma.yaml]: Already solved in #48756.

File: dts/bindings/dma/gd,gd32-dma.yaml
Comment (soburi) [reply dts/bindings/dma/gd,gd32-dma.yaml]: Already solved in #48756.

File: drivers/spi/spi_gd32.c
Comment (cameled) [inline drivers/spi/spi_gd32.c:325]: Can we move dma device check out of this function?

From below code, spi dma `TX` and `RX` enabled simultaneously.
```
	SPI_CTL1(cfg->reg) |= (SPI_CTL1_DMATEN | SPI_CTL1_DMAREN);
```

**Review by cameled — CHANGES_REQUESTED** [review]
Great work, some need to confirm.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c:400]: This check is necessary because DMA transfers may not be enabled for all SPIs even if DMA configuration enabled.

Comment (cameled) [reply drivers/spi/spi_gd32.c:400]: Suggestion:
```
#ifdef CONFIG_SPI_GD32_INTERRUPT
	if (IS_ENABLED(CONFIG_SPI_GD32_DMA) && spi_gd32_dma_enabled(dev)) {
		for (size_t i = 0; i < ARRAY_SIZE(data->dma); i++) {
			data->dma[i].count = 0;
		}

		ret = spi_gd32_start_dma_transceive(dev);
		if (ret < 0) {
			goto dma_error;
		}
		ret = spi_context_wait_for_completion(&data->ctx);
	} else {

		SPI_STAT(cfg->reg) &=
			~(SPI_STAT_RBNE | SPI_STAT_TBE | SPI_GD32_ERR_MASK);
		SPI_CTL1(cfg->reg) |=
			(SPI_CTL1_RBNEIE | SPI_CTL1_TBEIE | SPI_CTL1_ERRIE);
		ret = spi_context_wait_for_completion(&data->ctx);
	}
#else

	do {
		ret = spi_gd32_frame_exchange(dev);
		if (ret < 0) {
			break;
		}
	} while (spi_gd32_transfer_ongoing(data));

#ifdef CONFIG_SPI_ASYNC
	spi_context_complete(&data->ctx, dev, ret);
#endif
#endif
```

File: drivers/spi/spi_gd32.c
Comment (cameled) [inline drivers/spi/spi_gd32.c:369]: Shall we use `&&` in here? TX and RX dma can not work individually.

Comment (soburi) [reply drivers/spi/spi_gd32.c]: Change to require setting both TX and RX DMA channels.
Change to && here.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c]: Removed this check.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c]: Applied it.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c]: Change to declare as a global static variable.

Comment (soburi) [reply drivers/spi/spi_gd32.c]: Applied it with some modifications to solve compiler warnings.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c]: Applied it with some modifications to solve compiler warnings.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c]: DMA has been changed to guarantee its existence at initialization, so the check here is no longer necessary.
Removed from here.

**Review by nandojve — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_gd32.c
Comment (cameled) [reply drivers/spi/spi_gd32.c:92]: We should not access these from MCU, because DMA may access it at any time after spi driver initialized. 

SRAM can only be accessed by one AHB master at a time. So `dummy_tx` and `dummy_rx` can merge as one.

Comment (soburi) [reply drivers/spi/spi_gd32.c:92]: @nandojve 
Static variable initialize default with zero. So I think it is no problem.
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf §6.7.8/10
(It is rare to initialize a static variable explicitly with zero in zephyr.)

@cameled 
I want to fix the rx_dummy to zero.
Transmitting modifies the value if we merge tx_dummy and rx_dummy into a single variable.
I think it may make it harder to debug.

Comment (cameled) [reply drivers/spi/spi_gd32.c:92]: Hi @soburi , I update the above comment, as @nandojve suggested, we can give an explicit default zero to `dummy_tx`.

And you are right, `dummy_tx` and `dummy_rx` can not merge as one.

File: drivers/spi/spi_gd32.c
Comment (cameled) [inline drivers/spi/spi_gd32.c:264]: Suggestion: Remove `datareg`, asign `SPI_DATA` address directly.

File: drivers/spi/spi_gd32.c
Comment (cameled) [inline drivers/spi/spi_gd32.c:260]: Suggestion: Give `dir` a const, prevent unexpected change.

File: drivers/spi/spi_gd32.c
Comment (cameled) [inline drivers/spi/spi_gd32.c:257]: Suggestion: change `dmadev` to `dma_dev`,

**Review by cameled — COMMENTED** [review]
LGTM, add some suggestion.

File: drivers/spi/spi_gd32.c
Comment (cameled) [inline drivers/spi/spi_gd32.c:105]: Compared with `ARRAY_SIZE`, function call have an extra overhead.

File: drivers/spi/spi_gd32.c
Comment (nandojve) [reply drivers/spi/spi_gd32.c:92]: Fine for me, I just rose to make sure you folks prefer 0x00 or 0xFF.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c]: Applied your suggestion, Thank you.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c]: Applied your suggestion, Thank you.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c]: Applied your suggestion, Thank you.

File: drivers/spi/spi_gd32.c
Comment (soburi) [reply drivers/spi/spi_gd32.c:105]: Determining from ARRAY_SIZE won't work if both exists DMA enabled SPI and disabled one.

**Review by cameled — APPROVED** [review]

**Review by nandojve — APPROVED** [review]
Thank you @soburi !

**Review by teburd — APPROVED** [review]

**Review by fkokosinski — APPROVED** [review]

**Pull request merged**
